### PR TITLE
First pass at pipeline executions

### PR DESCRIPTION
**Pull request by robzienert**
Pretty naive implementation at this point, but it does work.

PTAL @cfieber @ajordens

### CODE REVIEW

File: orca-pipelinetemplate/src/test/resources/templates/simple-001.yml [hunk 1]
```diff
@@ -1,30 +1,36 @@
...
   name: Bake
   config:
-    regions: "{{ var.regions }}"
-    package: "{{ var.application }}-package"
+    regions: |
+      [
+      {{#each regions}}
+      "{{ this }}"{{#unless @last}},{{/unless}}
+      {{/each}}
+      ]
```
Comment (robzienert) [inline orca-pipelinetemplate/src/test/resources/templates/simple-001.yml]: One thing I hadn't really thought about until writing the integration spec is that handlebar templates need to emit  JSON, _not_ YAML. I'm not sure if we want to support YAML directly, or if telling people to emit JSON is fine. Might cause some room for confusion, though.

File: orca-pipelinetemplate/src/main/java/com/netflix/spinnaker/orca/pipelinetemplate/v1schema/V1SchemaExecutionGenerator.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        stage.put("context", s.getConfig());
+        stage.put("requisiteStageRefIds", getStageRequisiteIds(s, template.getStages()));
+        return stage;
+      })
+      .collect(Collectors.toList()));
+
+    return pipeline;
+  }
+
+  private static List<Integer> getStageRequisiteIds(StageDefinition stage, List<StageDefinition> allStages) {
```
Comment (robzienert) [inline orca-pipelinetemplate/src/main/java/com/netflix/spinnaker/orca/pipelinetemplate/v1schema/V1SchemaExecutionGenerator.java]: I've moved this into the graph transforms in a PR I'll be opening after this is merged.

File: orca-pipelinetemplate/src/main/java/com/netflix/spinnaker/orca/pipelinetemplate/v1schema/graph/transform/ConfigStageInjectionTransform.java [hunk 3]
```diff
@@ -74,6 +77,24 @@ private void replaceStages(PipelineTemplate pipelineTemplate) {
...
+
+    if (!configStages.isEmpty() || !templateStages.isEmpty()) {
+      throw new UnsupportedOperationException("stage injection is not yet supported");
+    }
+  }
+
+  private void assignRefIds(PipelineTemplate pipelineTemplate) {
+    AtomicInteger refId = new AtomicInteger(1);
+    pipelineTemplate.getStages().forEach(s -> {
+      s.setRefId(refId.getAndIncrement());
```
Comment (cfieber) [inline orca-pipelinetemplate/src/main/java/com/netflix/spinnaker/orca/pipelinetemplate/v1schema/graph/transform/ConfigStageInjectionTransform.java]: refId is actually a string in the JSON and just happens to be done as an incrementing integer in the UI. If there is already some other unique identifier that we have for each stage could probably just use that (and the resulting JSON maybe has some more useful meaning to it when looking at it). If we don't have an already unique identifier at this point I'm totally okay with just using ints

**Comment by cfieber** [comment]
LGTM

File: orca-pipelinetemplate/src/main/java/com/netflix/spinnaker/orca/pipelinetemplate/PipelineTemplatePipelinePreprocessor.java [hunk 4]
```diff
@@ -66,18 +68,18 @@ public PipelineTemplatePipelinePreprocessor(ObjectMapper pipelineTemplateObjectM
...
     PipelineTemplate template = TemplateMerge.merge(templates);
 
     GraphMutator graphMutator = new GraphMutator(templateConfiguration, renderer, registry);
     graphMutator.mutate(template);
 
-    // TODO final validation & marshal to pipeline json
+    // TODO validation
 
-    return pipeline;
+    ExecutionGenerator executionGenerator = new V1SchemaExecutionGenerator();
```
Comment (robfletcher) [inline orca-pipelinetemplate/src/main/java/com/netflix/spinnaker/orca/pipelinetemplate/PipelineTemplatePipelinePreprocessor.java]: Does the v1 here refer to orca's execution engine? Why do we want to run this stuff in v1?

File: orca-pipelinetemplate/src/main/java/com/netflix/spinnaker/orca/pipelinetemplate/v1schema/V1SchemaExecutionGenerator.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+public class V1SchemaExecutionGenerator implements ExecutionGenerator {
+
+  @Override
+  public Map<String, Object> generate(PipelineTemplate template, TemplateConfiguration configuration) {
+    Map<String, Object> pipeline = new HashMap<>();
+    pipeline.put("id", configuration.getRuntimeId());
+    pipeline.put("application", configuration.getPipeline().getApplication());
+    pipeline.put("name", Optional.ofNullable(configuration.getPipeline().getName()).orElse("Unnamed Execution"));
+    // TODO Doesn't look like we even use this anywhere in the codebase.
+//    pipeline.put("appConfig", "unsupported");
```
Comment (robfletcher) [inline orca-pipelinetemplate/src/main/java/com/netflix/spinnaker/orca/pipelinetemplate/v1schema/V1SchemaExecutionGenerator.java]: I _think_ this is for preferences set at an application level — notifications and the like

File: orca-pipelinetemplate/src/test/groovy/com/netflix/spinnaker/orca/pipelinetemplate/PipelineTemplatePipelinePreprocessorSpec.groovy [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+class PipelineTemplatePipelinePreprocessorSpec extends Specification {
+
+  ObjectMapper objectMapper = new ObjectMapper()
+
+  TemplateLoader templateLoader = new TemplateLoader([new FileTemplateSchemeLoader(objectMapper)])
+
+  Renderer renderer = new HandlebarsRenderer(objectMapper)
+
+  Registry registry = Mock() {
+    clock() >> Mock(Clock) {
```
Comment (robfletcher) [inline orca-pipelinetemplate/src/test/groovy/com/netflix/spinnaker/orca/pipelinetemplate/PipelineTemplatePipelinePreprocessorSpec.groovy]: What's this used for? I don't see anything referencing it elsewhere.

File: orca-pipelinetemplate/src/main/java/com/netflix/spinnaker/orca/pipelinetemplate/v1schema/graph/transform/ConfigStageInjectionTransform.java (hunk 3 above)
Comment (robzienert) [inline orca-pipelinetemplate/src/main/java/com/netflix/spinnaker/orca/pipelinetemplate/v1schema/graph/transform/ConfigStageInjectionTransform.java]: Oh fancy. If I change this from an integer value, do you think we'd wind up having render issues on the UI?

File: orca-pipelinetemplate/src/main/java/com/netflix/spinnaker/orca/pipelinetemplate/PipelineTemplatePipelinePreprocessor.java (hunk 4 above)
Comment (robzienert) [inline orca-pipelinetemplate/src/main/java/com/netflix/spinnaker/orca/pipelinetemplate/PipelineTemplatePipelinePreprocessor.java]: It references the pipeline template schema version, not the execution engine.

File: orca-pipelinetemplate/src/main/java/com/netflix/spinnaker/orca/pipelinetemplate/v1schema/V1SchemaExecutionGenerator.java (hunk 5 above)
Comment (robzienert) [inline orca-pipelinetemplate/src/main/java/com/netflix/spinnaker/orca/pipelinetemplate/v1schema/V1SchemaExecutionGenerator.java]: I was looking around in the codebase and didn't see it actually used anywhere of significance.

File: orca-pipelinetemplate/src/main/java/com/netflix/spinnaker/orca/pipelinetemplate/v1schema/graph/transform/ConfigStageInjectionTransform.java (hunk 3 above)
Comment (robfletcher) [inline orca-pipelinetemplate/src/main/java/com/netflix/spinnaker/orca/pipelinetemplate/v1schema/graph/transform/ConfigStageInjectionTransform.java]: Would be easy to try with an API call

File: orca-pipelinetemplate/src/test/groovy/com/netflix/spinnaker/orca/pipelinetemplate/PipelineTemplatePipelinePreprocessorSpec.groovy (hunk 6 above)
Comment (robzienert) [inline orca-pipelinetemplate/src/test/groovy/com/netflix/spinnaker/orca/pipelinetemplate/PipelineTemplatePipelinePreprocessorSpec.groovy]: There's a metric in the `PipelineTemplatePipelinePreprocessor` that records the elapsed time for processing templates into executions. This spec came sometime after that class was written.

File: orca-pipelinetemplate/src/main/java/com/netflix/spinnaker/orca/pipelinetemplate/v1schema/graph/transform/ConfigStageInjectionTransform.java (hunk 3 above)
Comment (robzienert) [inline orca-pipelinetemplate/src/main/java/com/netflix/spinnaker/orca/pipelinetemplate/v1schema/graph/transform/ConfigStageInjectionTransform.java]: Looks to work in the UI just fine! I'll update stuff. Thanks for the suggestion.

**Comment by robzienert** [comment]
@robfletcher Not the purpose of this PR. `PipelinePreprocessor` objects are [run before executions occur](https://github.com/spinnaker/orca/blob/master/orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/OperationsController.groovy#L67), and just are expected to result in pipeline execution object maps. This PR is just transforming the pipeline template into an execution.

If you like, I can give you a quick pass through the workflow.

**Pull request merged**